Using openexr via cmake add_subdirectory

I'm trying to use openexr as a statically build library in another project. I got it working by checking it out standalone, compiling it as static library and then putting this in my CMakeLists.txt

```
find_package(Imath REQUIRED HINTS ".../Software/openexr/my_manual_install")
find_package(OpenEXR REQUIRED HINTS ".../Software/openexr/my_manual_install")
include_directories( ".../Software/openexr/my_manual_install/include/Imath")
include_directories( ".../Software/openexr/my_manual_install/include/OpenEXR")
```

I would prefer to use it directly via add_subdirectory similar to this:
```
add_subdirectory( "source/Lib/Imath-3.1.12" )
add_subdirectory( "source/Lib/openexr-3.2.4" )
include_directories( "source/Lib/Imath-3.1.12/src/Imath") # this is wrong
include_directories( "source/Lib/openexr-3.2.4/src") # this is wrong
```
but was not successful. I'm having trouble finding the headers this way. Can't tell if linking would work as I'm didn't get that far. Reason I'm doing this that I'll need to build/run the project on other servers (cluster). My current approach should work as well, but it would be a much simpler workflow if I can get it to work with add_subdirectory, as I would not need to have a separate process to build it as a static library. I'm not sure if it can't be used that way or I'm simply doing it wrong. Any help appreciated!

Thank you very much for taking the time to have a look at this!

I've tried that approach in a simple test project based on sample code from here (https://openexr.com/en/latest/ReadingAndWritingImageFiles.html), but I run into this:
```
[cmake] CMake Error at openexr/cmake/CMakeLists.txt:159 (add_library):
[cmake]   add_library cannot create ALIAS target "MyModernCMakeExample::Config"
[cmake]   because another target with the same name already exists.
```

Played around a little, but I don't know enough cmake :/.  However, since you confirmed that OpenEXR is not supposed to be used like this I'll stick to the straight-forward approach: build it as a static library and then use that one.

Thanks again!

> Thank you very much for taking the time to have a look at this!非常感谢您花时间查看此内容！
> 
> I've tried that approach in a simple test project based on sample code from here (https://openexr.com/en/latest/ReadingAndWritingImageFiles.html), but I run into this:我已经在基于此处的示例代码（ https://openexr.com/en/latest/ReadingAndWritingImageFiles.html ）的简单测试项目中尝试了这种方法，但是遇到了这个问题：
> 
> ```
> [cmake] CMake Error at openexr/cmake/CMakeLists.txt:159 (add_library):
> [cmake]   add_library cannot create ALIAS target "MyModernCMakeExample::Config"
> [cmake]   because another target with the same name already exists.
> ```
> 
> Played around a little, but I don't know enough cmake :/. However, since you confirmed that OpenEXR is not supposed to be used like this I'll stick to the straight-forward approach: build it as a static library and then use that one.玩了一下，但我对 cmake 了解不够 :/。不过，既然你确认过 OpenEXR 不应该这样用，我就直接用直接的方法吧：先把它编译成静态库，然后再用那个。
> 
> Thanks again!  再次感谢！

I encountered a similar error(my `cmake` version is 3.26), so how can this problem be solved:

```CMake Error at src/ext/openexr/cmake/CMakeLists.txt:159 (add_library):
  add_library cannot create ALIAS target "OpenEXR::Config" because another
  target with the same name already exists.


CMake Error at src/ext/openexr/cmake/CMakeLists.txt:170 (add_library):
  add_library cannot create ALIAS target "OpenEXR::Config" because another
  target with the same name already exists.
```

I checked the relevant source code:
```###################################################
####### Interface libraries for configs

add_library(OpenEXRConfig INTERFACE)
target_include_directories(OpenEXRConfig INTERFACE
  $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
  $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
  $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/${OPENEXR_OUTPUT_SUBDIR}>)

if(OPENEXR_INSTALL)
  install(TARGETS OpenEXRConfig EXPORT ${PROJECT_NAME})
endif()
add_library(${PROJECT_NAME}::Config ALIAS OpenEXRConfig)

add_library(IexConfig INTERFACE)
target_include_directories(IexConfig INTERFACE
  $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
  $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
  $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/${OPENEXR_OUTPUT_SUBDIR}>)

if(OPENEXR_INSTALL)
  install(TARGETS IexConfig EXPORT ${PROJECT_NAME})
endif()
add_library(${PROJECT_NAME}::Config ALIAS IexConfig)

add_library(IlmThreadConfig INTERFACE)
target_include_directories(IlmThreadConfig INTERFACE
  $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
  $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
  $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/${OPENEXR_OUTPUT_SUBDIR}>)

if(OPENEXR_INSTALL)
  install(TARGETS IlmThreadConfig EXPORT ${PROJECT_NAME})
endif()
add_library(${PROJECT_NAME}::Config ALIAS IlmThreadConfig)
```

when i use 3.15, the error gone.

It is recommended to modify cmake to maintain compatibility.

Please let me explain in detail what happened.
I use `add_subdirectory` to use `openexr` as a submodule and use the following command to build it:
```
cmake -B build/debug -DCMAKE_BUILD_TYPE=Debug -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_POLICY_VERSION_MINIMUM=3.26
cmake --build build/debug --parallel 4
```
cmake will report an error during the configuration phase:
```
CMake Error at src/ext/openexr/cmake/CMakeLists.txt:159 (add_library):
  add_library cannot create ALIAS target "OpenEXR::Config" because another
  target with the same name already exists.


CMake Error at src/ext/openexr/cmake/CMakeLists.txt:170 (add_library):
  add_library cannot create ALIAS target "OpenEXR::Config" because another
  target with the same name already exists.
```

when i use:
```
cmake -B build/debug -DCMAKE_BUILD_TYPE=Debug -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_POLICY_VERSION_MINIMUM=3.15
```
 the error gone.

I'm not quite sure what changes have been made to the higher version of cmake. A reasonable speculation is that it no longer allows the following code to be written in the same way:
```
####### Interface libraries for configs

add_library(OpenEXRConfig INTERFACE)
target_include_directories(OpenEXRConfig INTERFACE
  $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
  $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
  $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/${OPENEXR_OUTPUT_SUBDIR}>)

if(OPENEXR_INSTALL)
  install(TARGETS OpenEXRConfig EXPORT ${PROJECT_NAME})
endif()
add_library(${PROJECT_NAME}::Config ALIAS OpenEXRConfig)

add_library(IexConfig INTERFACE)
target_include_directories(IexConfig INTERFACE
  $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
  $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
  $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/${OPENEXR_OUTPUT_SUBDIR}>)

if(OPENEXR_INSTALL)
  install(TARGETS IexConfig EXPORT ${PROJECT_NAME})
endif()
add_library(${PROJECT_NAME}::Config ALIAS IexConfig)

add_library(IlmThreadConfig INTERFACE)
target_include_directories(IlmThreadConfig INTERFACE
  $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
  $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
  $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/${OPENEXR_OUTPUT_SUBDIR}>)

if(OPENEXR_INSTALL)
  install(TARGETS IlmThreadConfig EXPORT ${PROJECT_NAME})
endif()
add_library(${PROJECT_NAME}::Config ALIAS IlmThreadConfig)
```

I would recommend modifying the cmake configuration related to openexr to maintain compatibility with higher versions of cmake, as we need to keep up with the times, right

--- a/cmake/CMakeLists.txt
+++ b/cmake/CMakeLists.txt
@@ -156,7 +156,7 @@ target_include_directories(IexConfig INTERFACE
 if(OPENEXR_INSTALL)
   install(TARGETS IexConfig EXPORT ${PROJECT_NAME})
 endif()
-add_library(${PROJECT_NAME}::Config ALIAS IexConfig)
+add_library(Iex::Config ALIAS IexConfig)
 
 add_library(IlmThreadConfig INTERFACE)
 target_include_directories(IlmThreadConfig INTERFACE
@@ -167,7 +167,7 @@ target_include_directories(IlmThreadConfig INTERFACE
 if(OPENEXR_INSTALL)
   install(TARGETS IlmThreadConfig EXPORT ${PROJECT_NAME})
 endif()
-add_library(${PROJECT_NAME}::Config ALIAS IlmThreadConfig)
+add_library(IlmThread::Config ALIAS IlmThreadConfig)
 
 ###################################################
 ####### Install pkg-config files if requested

